examples/cgroup-rstat-flushing.bpf.c

```diff
@@ -0,0 +1,211 @@
+    u32 bucket;
+};
+
+struct {
+    __uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+    __uint(max_entries, MAX_LATENCY_SLOT + 1);
+    __type(key, struct hist_key_t);
+    __type(value, u64);
+} cgroup_rstat_lock_wait_seconds SEC(".maps");
+
+struct {
+    __uint(type, BPF_MAP_TYPE_LRU_HASH);
+    __uint(max_entries, 1024 * 128);
+    __type(key, u64);
+    __type(value, u64);
+} start_wait SEC(".maps");
+
+struct {
+    __uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+    __uint(max_entries, MAX_LATENCY_SLOT + 1);
+    __type(key, struct hist_key_t);
+    __type(value, u64);
+} cgroup_rstat_lock_hold_seconds SEC(".maps");
+
+struct {
+    __uint(type, BPF_MAP_TYPE_LRU_HASH);
+    __uint(max_entries, 1024 * 128);
+    __type(key, u64);
+    __type(value, u64);
+} start_hold SEC(".maps");
+
+/* Complex key for encoding lock properties */
+struct lock_key_t {
+	u8  contended;
+	u8  yield;
+	u16 level;
+};
+#define MAX_LOCK_KEY_ENTRIES	128
+
+/* Total counter for obtaining lock together with state (prometheus labels)
+ *
+ * State for cgroup level, contended and yield case.
+ *
+ * The problematic/interesting case is when the lock was contended (prior to
+ * obtaining the lock).  This "contended" label is key in analyzing locking
+ * issues.
+ *
+ * Kernel can yield the rstat lock when walking individial CPU stats.  This
+ * leads to "interresting" concurrency issues.  Thus, having a label "yield"
+ * can help diagnose.
+ */
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_HASH);
+	__uint(max_entries, MAX_LOCK_KEY_ENTRIES);
+	__type(key, struct lock_key_t);
+	__type(value, u64);
+} cgroup_rstat_locked_total SEC(".maps");
+
+/** Measurement#1: lock rates
+ *  =========================
+ For locks, the problematic/interesting case is when the lock was contended.
+
+ Simply counting the lock+unlock is complicated by yielding the lock in the main
+ flushing loop (in cgroup_rstat_flush_locked()). The tracepoints "cpu" argument
+ will be (minus) -1 when the lock is not a yielded lock.
+
+ Concern: The lock rates will vary (a lot) and aggregating this as an average
+ will not capture spikes. Especially given Prometheus capture intervals only
+ happens every 53 seconds.
+
+*/
+SEC("tp_btf/cgroup_rstat_locked")
+int BPF_PROG(rstat_locked, struct cgroup *cgrp, int cpu, bool contended)
+{
+	u64 now = bpf_ktime_get_ns();
+	u64 pid = bpf_get_current_pid_tgid();
+	struct lock_key_t lock_key = { 0 };
+	u32 level = cgrp->level;
+
+	if (level > MAX_CGROUP_LEVELS)
+		level = MAX_CGROUP_LEVELS;
+
+	if (cpu >= 0)
+		lock_key.yield = 1;
+
+	lock_key.contended = contended;
+	lock_key.level = (level & 0xFF);
+
+	increment_map_nosync(&cgroup_rstat_locked_total, &lock_key, 1);
+
+	/* Lock hold time start */
+	bpf_map_update_elem(&start_hold, &pid, &now, BPF_ANY);
+	// TODO: Should we ignore yield and measure flush time instead?
+
+	/* Lock contended event happened prior to obtaining this lock.
+	 * Get back start "wait" timestamp that was recorded.
+	 */
+	if (contended) {
```

Comment:
This can be an early return.

Reply:
True for current code, but I might add more code later.
And I think the code is more readable IMHO, as it is more clear that everything under `if (contended) {` is handling the contended case.